gtest/src/system.rs

```diff
@@ -62,10 +66,53 @@ impl System {
         self.0.borrow_mut().block_info.timestamp += amount as u64;
     }
 
-    pub fn get_program<ID: Into<ProgramIdWrapper>>(&'_ self, id: ID) -> Program<'_> {
-        Program {
-            manager: &self.0,
-            id: id.into().0,
-        }
+    /// Gets a program by `id`
+    ///
+    /// Returns `None` if program with `id` doesn't exist or is a dormant actor.
+    pub fn get_program<ID: Into<ProgramIdWrapper>>(&'_ self, id: ID) -> Option<Program<'_>> {
+        let id = id.into().0;
+        self.0
+            .borrow()
+            .actors
+            .get(&id)
+            .and_then(|(actor, _)| match actor {
+                Actor::Dormant => None,
+                _ => Some(Program {
+                    id,
+                    manager: &self.0,
+                }),
+            })
+    }
+
+    /// Returns program ids of successfully initialized programs.
+    ///
+    /// Ids of uninitialized and failed during initialization programs aren't returned.
+    pub fn initialized_programs(&self) -> Vec<ProgramId> {
+        self.0
+            .borrow()
+            .actors
+            .iter()
+            .filter_map(|(&program_id, (actor, _))| match actor {
+                Actor::Initialized(_) => Some(program_id),
+                _ => None,
+            })
+            .collect()
+    }
```

Comment:
why do we need this method? and can't we replace it with `system.program_active(id) -> bool` ?

Reply:
Actually needed that for the case, when someone tries to create a duplicate program from program. In [de9c434](https://github.com/gear-tech/gear/pull/755/commits/de9c43448972814931501a63dbbcf731887b7ec4) I have introduced `total_processed` messages counter. Now duplicate check can be done by observing amount of messages processed.